modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/infrastructure/consensus/ValidationErrorStorage.scala

```diff
@@ -0,0 +1,54 @@
+package io.constellationnetwork.node.shared.infrastructure.consensus
+
+import java.time.Instant
+
+import cats.data.NonEmptyList
+import cats.effect.kernel.{Async, Clock, Ref}
+import cats.effect.std.Semaphore
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+trait ValidationErrorStorage[F[_], Artifact, ValidationError] {
+  def addValidationError(artifact: Artifact, errors: ValidationError): F[Unit]
+  def addValidationErrors(errors: NonEmptyList[(Artifact, ValidationError)]): F[Unit]
+  def allValidationErrors: F[Vector[(Artifact, ValidationError, Instant)]]
+}
+
+object ValidationErrorStorage {
+
+  def make[F[_]: Async, Artifact, ValidationError](
+    maxSize: Int
+  )(implicit clock: Clock[F]): F[ValidationErrorStorage[F, Artifact, ValidationError]] =
+    for {
+      updateSemaphore <- Semaphore[F](1)
```

Comment:
Why do you need a semaphore, what was your motivation to use it together with `Ref`? `Ref` is thread-safe and supports concurrent read/write (we use it a lot in the codebase). Is there anything special that it needs to be additionally locked?